Mathlib/FieldTheory/SeparableDegree.lean

```diff
@@ -268,31 +249,539 @@ def embProdEmbOfIsAlgebraic [Algebra E K] [IsScalarTower F E K] (halg : Algebra.
+with arbitrary large finite extension degree. -/
+lemma exists_lt_finrank_of_not_finiteDimensional
+    (halg : Algebra.IsAlgebraic F E) (hnfd : ¬ FiniteDimensional F E) (n : ℕ) :
+    ∃ L : IntermediateField F E, FiniteDimensional F L ∧ n < finrank F L := by
+  induction' n with n ih
+  · exact ⟨⊥, Subalgebra.finiteDimensional_bot, by
+      rw [IntermediateField.finrank_bot]; exact one_pos⟩
+  obtain ⟨L, _, hn⟩ := ih
+  by_cases h : n + 1 < finrank F L
+  · exact ⟨L, by assumption, h⟩
+  have hr := eq_of_ge_of_not_gt (by exact hn) h
+  obtain ⟨x, hx⟩ := show ∃ x : E, x ∉ L by
+    by_contra hx
+    simp only [not_exists, not_not] at hx
+    have : L = ⊤ := eq_top_iff.2 fun x _ ↦ hx x
+    erw [this, finrank_top F E, finrank_of_infinite_dimensional hnfd] at hr
+    linarith only [hr]
+  let L' := L ⊔ F⟮x⟯
+  haveI := adjoin.finiteDimensional (halg x).isIntegral
+  haveI := finiteDimensional_sup L F⟮x⟯
+  refine ⟨L', by assumption, ?_⟩
+  have : finrank F L < finrank F L' := by
+    by_contra h
+    have h1 : L = L' := eq_of_le_of_finrank_le le_sup_left (not_lt.1 h)
+    have h2 : F⟮x⟯ ≤ L' := le_sup_right
+    exact hx <| (h1.symm ▸ h2) <| mem_adjoin_simple_self F x
+  rwa [hr] at this
+
+end IntermediateField
+
+namespace Field
+
+/-- The separable degree of any field extension `E / F` divides the degree of `E / F`. -/
+theorem finSepDegree_dvd_finrank : finSepDegree F E ∣ finrank F E := by
+  by_cases hfd : FiniteDimensional F E
+  · let P : IntermediateField F E → Prop := fun K ↦ finSepDegree F K ∣ finrank F K
+    rw [← finSepDegree_top F, ← finrank_top F E]
+    refine induction_on_adjoin P ?_ (fun L x h ↦ ?_) ⊤
+    · simp only [finSepDegree_bot, IntermediateField.finrank_bot, one_dvd]
+    simp only at h ⊢
+    have hdvd := mul_dvd_mul h <| finSepDegree_adjoin_simple_dvd_finrank L E x
+    set M := L⟮x⟯; clear_value M
+    letI : Algebra L M := Subalgebra.algebra M.toSubalgebra
+    letI : Module L M := Algebra.toModule
+    letI : SMul L M := Algebra.toSMul
+    haveI : IsScalarTower F L M := IntermediateField.isScalarTower M
+    rwa [finSepDegree_mul_finSepDegree_of_isAlgebraic F L M (Algebra.IsAlgebraic.of_finite L M),
+      FiniteDimensional.finrank_mul_finrank F L M] at hdvd
+  rw [finrank_of_infinite_dimensional hfd]
+  exact dvd_zero _
+
+/-- The separable degree of a finite extension `E / F` is smaller than the degree of `E / F`. -/
+theorem finSepDegree_le_finrank [FiniteDimensional F E] :
+    finSepDegree F E ≤ finrank F E := Nat.le_of_dvd finrank_pos <| finSepDegree_dvd_finrank F E
+
+/-- If `E / F` is a separable extension, then its separable degree is equal to its degree.
+When `E / F` is infinite, it means that `Field.Emb F E` has infinitely many elements.
+(But the cardinality of `Field.Emb F E` is not equal to `Module.rank F E` in general!) -/
+theorem finSepDegree_eq_finrank_of_isSeparable [IsSeparable F E] :
+    finSepDegree F E = finrank F E := by
+  clear! K
```

Comment:
Why is this line needed?

Reply:
Without this the following `wlog` becomes buggy (I think).

Reply:
Can you add a comment explaining this?

Reply:
Sure.